### ISSUE TITLE
Null default for empty collections is annoying

### ISSUE THREAD

**Issue by stevenschlansker** [google/auto#451]
It's very annoying to have to write:

```java
    @AutoValue.Builder
    public abstract static class Builder {
        public static Builder create() {
            return new AutoValue_MyType.Builder()
                    .setCharges(emptyList())
                    .setPayments(emptyList())
                    .setParticipants(emptyList())
                    .setTags(emptyList());
        }
```

just to ensure that a field `@Nonnull public abstract List<String> getCharges()` defaults to empty-list on the builder instead of null.  It would be nice to have some way to specify a default for non-primitive non-null types.

**Comment by hzsweers** [comment]
you can specify default values in your builder implementation

**Comment by lowasser** [comment]
@hzsweers, it looks like that's exactly what the OP is already doing.

I think the complication here is that this means AutoValue would have to pick a default _implementation_ of `List`.  I think there was a discussion in another issue recently that this was easy enough for e.g. `ImmutableList`, but `List` -- and even more, other collection implementations -- makes it much less clear.

**Comment by hzsweers** [comment]
It read to me like a general ask for default values support

**Comment by eamonnmcmanus** [comment]
Yeah, I supposed it was asking for something maybe like this:

```java
@AutoValue
public abstract class Foo {
  public abstract List<String> bars();
  public abstract List<String> bazzes();
  public static Builder builder() {
    return new AutoValue_Foo.Builder();
  }

  @AutoValue.Builder(
    defaults = {@AutoValue.Default(List.class, "java.util.Collections.emptyList()")})
  public abstract class Builder {
    public abstract Builder setBars(List<String> x);
    public abstract Builder setBazzes(List<String> x);
    public abstract Foo build();
  }
}
```

In the original example there are four List properties, each of which has to be defaulted to Collections.emptyList(). With an incantation like the above, they would all automatically get the same default value.

I'm not really convinced by this, though. It's a lot of mechanism with a lot of complicated detail (what about type parameters, for example), and I don't think it is really all that burdensome to have to spell out the individual defaults explicitly.

I'll note too that we recommend using ImmutableList rather than List, and if in addition you have `public abstract ImmutableList.Builder<String> chargesBuilder()` then the default will indeed be the empty list.

**Comment by stevenschlansker** [comment]
Yes, I agree, the annotation is a little ugly and verbose for what I was asking... maybe there isn't a good solution and this just exists as a wart, but essentially if you have a lot of "collecting" types on a build it is very frustrating that the default behavior of `FooType.builder().build()` is to throw NPE for every collection type you don't have any of.

It is cool that you can have builders themselves have `ImmutableList.Builder` methods, but that probably doesn't change the behavior when you don't call `chargesBuilder()` at all -- I guess it is NPE, not empty list, right?

After thinking about this more, maybe this is a general case of a "collectable" type, as in `Collector<T>`.  Wouldn't it be cool if you could add builder methods like

```java
@AutoValue.Builder
class FooBuilder {
    Builder setBars(List<String> bar);
    @AutoValue.Collector(Collectors.toList())
    Builder addBar(String bar);
}
```

(Syntax is obviously not usable as-is, but hopefully you get the idea)

**Comment by eamonnmcmanus** [comment]
In fact, if you don't call `chargesBuilder()` at all, you do get an empty list.

The idea of reusing Collectors to simplify certain operations is interesting and worth thinking more about.
